src/index.d.ts

```diff
@@ -0,0 +1,72 @@
+export default Renderer;
+export { createElement, generateHTML, safe, htmlEncode };
+
+// ** TSX configuration
+
+declare global {
+	namespace JSX {
+		// This is needed for attribute type checking in custom components
+		interface Element extends StatelessFunctionalComponent<any> { }
+	}
+}
+
+// ** Types and interfaces
+
+// Return value of createElement
+declare type elementGenerator = (
+	stream: Renderer.Stream,
+	nonBlocking: boolean,
+	callback: () => void
+) => void;
+
+// A macro
+interface StatelessFunctionalComponent<T> {
+	(props: T): elementGenerator;
+}
+
+// ** Exports
+
+declare function createElement<T>(
+	element: string | StatelessFunctionalComponent<T>,
+	params: T,
+	...children
+): elementGenerator;
+
+declare function generateHTML(
+	tag: string,
+	params,
+	...children
+): elementGenerator;
+
+declare function safe(str: string): Renderer.HTMLString;
+
+declare function htmlEncode(str: string, attribute: boolean): string;
+
+declare class Renderer {
+	constructor(doctype?: string);
+
+	renderView(
+		view: elementGenerator | string,
+		params: Object,
+		stream: Renderer.Stream,
+		{ fragment: boolean }?,
+		callback?: () => void
+	): void;
+
+	registerView(
+		macro: () => elementGenerator,
+		name?: string,
+		replace?: boolean
+	): string;
+}
+
+declare namespace Renderer {
+	export interface Stream {
+		write(msg: string): void;
+		writeln(msg: string): void;
+		flush(): void;
+	}
+	export class HTMLString {
```

Comment:
Didn’t we mean to declare a supertype of `string` and `HTMLString` for convenience?

I believe something like
```ts
type String = string | Renderer.HTMLString
```
should work.

Reply:
Except then users would have to employ `String` rather than `string` - not ideal?

Reply:
Well an `HTMLString` _is_ not a `string`, it’s something distinct that we know to extract a string _from_.  There’s special handling for them in complate-stream but that doesn’t mean users can just use either interchangeably anywhere!

So as it is if a user wants to accept both `string` and `HTMLString` they’d have to declare a type of `string | HTMLString`, which is more verbose than `String`.

If we want users to be able to use `string` to also accept `HTMLString`s, we’d need to ensure that an `HTMLString` _is a_ `string` first.

Reply:
On second thought let’s discuss this elsewhere. Affordances for potential use cases don’t belong in this pull request, we need these basics first.

Reply:
btw: There already is a `String` class: https://developer.mozilla.org/de/docs/Web/JavaScript/Reference/Global_Objects/String :)

Reply:
Actually, why are `Stream` and `HTMLString` declared in the `Renderer` namespace?

`Renderer` is just a convenience wrapper on top of complate-stream's core functionality. Both the stream interface and `HTMLString` originate in the `html.js` module.